### ISSUE TITLE
Problem about inferencing on custom data

### ISSUE THREAD

**Issue by winterfell2021** [liqi-casia/FaceSwapper#5]
Hi, there, thank for your great work!
I write a simple demo about inferencing on custom data without generate landmark or mask images, the only difference is that i implement a function named `get_mark_parsing` to generate landmark and parse images. But i got a bad result with original inference. If u have time, plz give me some suggestion about where im wrong, Thx.
Full code as following:

```python
from core.model import build_model
from core import utils
from core.checkpoint import CheckpointIO
from torchvision import transforms
from PIL import Image
import torch.nn.functional as F
import torchvision.utils as vutils
import numpy as np


from parse import BiSeNet
import torch
n_classes = 19
net = BiSeNet(n_classes=n_classes)
net.cuda()
net.load_state_dict(torch.load('./pretrained_checkpoints/79999_iter.pth'))
net.eval()


config = dict(img_size=256,id_dim=512,wing_path='pretrained_checkpoints/wing.ckpt')
nets, nets_ema = build_model(config)
for name, module in nets.items():
    nets[name].to('cuda')
    if ('ema' not in name) and ('fan' not in name) and ('arcface' not in name):
        print('Initializing %s...' % name)
        module.apply(utils.he_init)
for name, module in nets_ema.items():
    nets_ema[name].to('cuda')
ckptios = [CheckpointIO('pretrained_checkpoints/', **nets_ema)]
for ckptio in ckptios:
    ckptio.load_test('faceswapper.ckpt')
    
    
src_dir = 'data/FF++_Dataset/ff++/raw_003_0.png'
tar_dir = src_dir.replace('raw_003_0.png', 'raw_000_0.png')

to_tensor = transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.485, 0.456, 0.406), (0.229, 0.224, 0.225)),
])

img_size = 256
transform = transforms.Compose([
        transforms.Resize([img_size, img_size]),
        transforms.ToTensor(),
        transforms.Normalize(mean=[0.5, 0.5, 0.5],
                             std=[0.5, 0.5, 0.5]),
])
transform_seg = transforms.Compose([
        transforms.Resize([img_size, img_size]),
        transforms.ToTensor(),
])

def get_mark_parsing(src_img):
    image = src_img.resize((512, 512), Image.BILINEAR)
    image = to_tensor(image)
    image = torch.unsqueeze(image, 0).cuda()
    out = net(image)[0]
    
    parsing = out.squeeze(0).cpu().detach().numpy().argmax(0)
    
    mark = parsing.copy()
    index = np.where(((parsing>0) & (parsing<=6)) | ((parsing>= 10) & (parsing<=13)))
    mark[index[0], index[1]] = 255
    index = np.where(mark!=255)
    mark[index[0], index[1]] = 0
    mark = Image.fromarray(np.uint8(mark))
    
    parsing = Image.fromarray(np.uint8(parsing))
    return mark, parsing

source_image = Image.open(src_dir).convert('RGB')
source_mask_image, source_parsing = get_mark_parsing(source_image)

tar_image = Image.open(tar_dir).convert('RGB')
tar_mask_image, tar_parsing = get_mark_parsing(tar_image)

source_lm_image = transform(source_image)
source_lm_image = nets_ema.fan.get_landmark_curve(source_lm_image.to('cuda').unsqueeze(0))
source_lm_image = source_lm_image[0]
source_lm_image = Image.fromarray(source_lm_image)

tar_lm_image = transform(tar_image)
tar_lm_image = nets_ema.fan.get_landmark_curve(tar_lm_image.to('cuda').unsqueeze(0))
tar_lm_image = tar_lm_image[0]
tar_lm_image = Image.fromarray(tar_lm_image)

source_image = transform(source_image)
source_lm_image = transform(source_lm_image)
source_mask_image = transform_seg(source_mask_image)
source_parsing = transform_seg(source_parsing)

tar_image = transform(tar_image)
tar_lm_image = transform(tar_lm_image)
tar_mask_image = transform_seg(tar_mask_image)
tar_parsing_image = transform_seg(tar_parsing)

src=source_image.to('cuda')
tar=tar_image.to('cuda')
src_lm=source_lm_image.to('cuda')
tar_lm=tar_lm_image.to('cuda')
src_mask=(1-source_mask_image).to('cuda')
tar_mask=(1-tar_mask_image).to('cuda')
src_parsing=source_parsing.to('cuda')
tar_parsing=tar_parsing_image.to('cuda')

srcid_taratt, tarid_srcatt,_,_,_,_ = nets_ema.generator(src.unsqueeze(dim=0), 
                                                    tar.unsqueeze(dim=0), 
                                                    src_lm.unsqueeze(dim=0),
                                                    tar_lm.unsqueeze(dim=0), 
                                                    src_mask.unsqueeze(dim=0), 
                                                    tar_mask.unsqueeze(dim=0))
def denormalize(x):
    out = (x + 1) / 2
    return out.clamp_(0, 1)
x = srcid_taratt[0,:,:,:]
x = denormalize(x)
vutils.save_image(x.cpu(), 'test.jpg', nrow=1, padding=0)
```

**Comment by liqi-casia** [comment]
`source_lm_image = transform(source_lm_image)` shoud be `source_lm_image = transform_seg(source_lm_image)`,
and  `tar_lm_image = transform(tar_lm_image)` shoud be `tar_lm_image = transform_seg(tar_lm_image)`

**Comment by ghost** [comment]
Unfortunately, there is no full-fledged demo from the developer, so I decided to test the code that was posted above... It seems that with it you can only create dudes thar all will be similar to Sidorovich from this screenshot)
![test](https://user-images.githubusercontent.com/63778828/163733332-9e5d2446-c023-4268-9417-fc97131861b7.jpg)
![images](https://user-images.githubusercontent.com/63778828/163733331-216414d7-2b80-49ad-baca-6e0f2e5d8209.jpeg)

**Comment by winterfell2021** [comment]
> Unfortunately, there is no full-fledged demo from the developer, so I decided to test the code that was posted above... It seems that with it you can only create dudes thar all will be similar to Sidorovich from this screenshot) ![test](https://user-images.githubusercontent.com/63778828/163733332-9e5d2446-c023-4268-9417-fc97131861b7.jpg) ![images](https://user-images.githubusercontent.com/63778828/163733331-216414d7-2b80-49ad-baca-6e0f2e5d8209.jpeg)

After patch the code from authors, the result changed from
![image](https://user-images.githubusercontent.com/85819194/163742334-706ede09-a837-4571-998e-4235c5f6089b.png)
to
![image](https://user-images.githubusercontent.com/85819194/163742387-4dcce2c0-2739-4400-8a6b-d393de55062e.png)

But still work bad with own data as your shows.

**Comment by liqi-casia** [comment]
1. For other datasets, please refer to the paragraph **Other Datasets** ( especially _First, crop and align face images from other datasets automatically so that the proportion of face occupied in the whole is similar to that of CelebA dataset and FaceForensics++ dataset._ )
2. _work bad_ you mean obvious artifacts or obsecure result? We found face image resolution in ff++ is ususlly low.
For example, you can see the original target image raw_000_0.png.

**Comment by winterfell2021** [comment]
> 2\. _work bad_ you mean obvious artifacts or obsecure result

![image](https://user-images.githubusercontent.com/85819194/163752483-0c8a68ae-6444-44e3-9aac-dc2372003598.png)
A sample swapping result on dfdc 2022 creation task, plz help where i mistake, as using the above patch.

**Comment by liqi-casia** [comment]
The possible reasons are listed as follows.
1. Our swapping model is trained on part of CelebA dataset (only 30,000 images). Most of them are caucasian. Besides, we don't utilize any video images. May be you should retrain our method.
2. As far as I know, there are no groudtruth images for face swapping. Thus you can finetune our model on dfdc 2022 dataset  if there is no obvious regulation against this.

**Comment by winterfell2021** [comment]
> The possible reasons are listed as follows.
> 
> 1. Our swapping model is trained on part of CelebA dataset (only 30,000 images). Most of them are caucasian. Besides, we don't utilize any video images. May be you should retrain our method.
> 2. As far as I know, there are no groudtruth images for face swapping. Thus you can finetune our model on dfdc 2022 dataset  if there is no obvious regulation against this.

Thanks for your kind suggestion, I will try to train with different dataset to find if asian-only-dataset improve its performance.

**Comment by cwalt2014** [comment]
> Hi, there, thank for your great work! I write a simple demo about inferencing on custom data without generate landmark or mask images, the only difference is that i implement a function named `get_mark_parsing` to generate landmark and parse images. But i got a bad result with original inference. If u have time, plz give me some suggestion about where im wrong, Thx. Full code as following:
> 
> ```python
> from core.model import build_model
> from core import utils
> from core.checkpoint import CheckpointIO
> from torchvision import transforms
> from PIL import Image
> import torch.nn.functional as F
> import torchvision.utils as vutils
> import numpy as np
> 
> 
> from parse import BiSeNet
> import torch
> n_classes = 19
> net = BiSeNet(n_classes=n_classes)
> net.cuda()
> net.load_state_dict(torch.load('./pretrained_checkpoints/79999_iter.pth'))
> net.eval()
> 
> 
> config = dict(img_size=256,id_dim=512,wing_path='pretrained_checkpoints/wing.ckpt')
> nets, nets_ema = build_model(config)
> for name, module in nets.items():
>     nets[name].to('cuda')
>     if ('ema' not in name) and ('fan' not in name) and ('arcface' not in name):
>         print('Initializing %s...' % name)
>         module.apply(utils.he_init)
> for name, module in nets_ema.items():
>     nets_ema[name].to('cuda')
> ckptios = [CheckpointIO('pretrained_checkpoints/', **nets_ema)]
> for ckptio in ckptios:
>     ckptio.load_test('faceswapper.ckpt')
>     
>     
> src_dir = 'data/FF++_Dataset/ff++/raw_003_0.png'
> tar_dir = src_dir.replace('raw_003_0.png', 'raw_000_0.png')
> 
> to_tensor = transforms.Compose([
>         transforms.ToTensor(),
>         transforms.Normalize((0.485, 0.456, 0.406), (0.229, 0.224, 0.225)),
> ])
> 
> img_size = 256
> transform = transforms.Compose([
>         transforms.Resize([img_size, img_size]),
>         transforms.ToTensor(),
>         transforms.Normalize(mean=[0.5, 0.5, 0.5],
>                              std=[0.5, 0.5, 0.5]),
> ])
> transform_seg = transforms.Compose([
>         transforms.Resize([img_size, img_size]),
>         transforms.ToTensor(),
> ])
> 
> def get_mark_parsing(src_img):
>     image = src_img.resize((512, 512), Image.BILINEAR)
>     image = to_tensor(image)
>     image = torch.unsqueeze(image, 0).cuda()
>     out = net(image)[0]
>     
>     parsing = out.squeeze(0).cpu().detach().numpy().argmax(0)
>     
>     mark = parsing.copy()
>     index = np.where(((parsing>0) & (parsing<=6)) | ((parsing>= 10) & (parsing<=13)))
>     mark[index[0], index[1]] = 255
>     index = np.where(mark!=255)
>     mark[index[0], index[1]] = 0
>     mark = Image.fromarray(np.uint8(mark))
>     
>     parsing = Image.fromarray(np.uint8(parsing))
>     return mark, parsing
> 
> source_image = Image.open(src_dir).convert('RGB')
> source_mask_image, source_parsing = get_mark_parsing(source_image)
> 
> tar_image = Image.open(tar_dir).convert('RGB')
> tar_mask_image, tar_parsing = get_mark_parsing(tar_image)
> 
> source_lm_image = transform(source_image)
> source_lm_image = nets_ema.fan.get_landmark_curve(source_lm_image.to('cuda').unsqueeze(0))
> source_lm_image = source_lm_image[0]
> source_lm_image = Image.fromarray(source_lm_image)
> 
> tar_lm_image = transform(tar_image)
> tar_lm_image = nets_ema.fan.get_landmark_curve(tar_lm_image.to('cuda').unsqueeze(0))
> tar_lm_image = tar_lm_image[0]
> tar_lm_image = Image.fromarray(tar_lm_image)
> 
> source_image = transform(source_image)
> source_lm_image = transform(source_lm_image)
> source_mask_image = transform_seg(source_mask_image)
> source_parsing = transform_seg(source_parsing)
> 
> tar_image = transform(tar_image)
> tar_lm_image = transform(tar_lm_image)
> tar_mask_image = transform_seg(tar_mask_image)
> tar_parsing_image = transform_seg(tar_parsing)
> 
> src=source_image.to('cuda')
> tar=tar_image.to('cuda')
> src_lm=source_lm_image.to('cuda')
> tar_lm=tar_lm_image.to('cuda')
> src_mask=(1-source_mask_image).to('cuda')
> tar_mask=(1-tar_mask_image).to('cuda')
> src_parsing=source_parsing.to('cuda')
> tar_parsing=tar_parsing_image.to('cuda')
> 
> srcid_taratt, tarid_srcatt,_,_,_,_ = nets_ema.generator(src.unsqueeze(dim=0), 
>                                                     tar.unsqueeze(dim=0), 
>                                                     src_lm.unsqueeze(dim=0),
>                                                     tar_lm.unsqueeze(dim=0), 
>                                                     src_mask.unsqueeze(dim=0), 
>                                                     tar_mask.unsqueeze(dim=0))
> def denormalize(x):
>     out = (x + 1) / 2
>     return out.clamp_(0, 1)
> x = srcid_taratt[0,:,:,:]
> x = denormalize(x)
> vutils.save_image(x.cpu(), 'test.jpg', nrow=1, padding=0)
> ```

Can you post your code for get_mark_parsing() ?

**Comment by ghost** [comment]
> Hi, there, thank for your great work! I write a simple demo about inferencing on custom data without generate landmark or mask images, the only difference is that i implement a function named `get_mark_parsing` to generate landmark and parse images. But i got a bad result with original inference. If u have time, plz give me some suggestion about where im wrong, Thx. Full code as following:
> 
> ```python
> from core.model import build_model
> from core import utils
> from core.checkpoint import CheckpointIO
> from torchvision import transforms
> from PIL import Image
> import torch.nn.functional as F
> import torchvision.utils as vutils
> import numpy as np
> 
> 
> from parse import BiSeNet
> import torch
> n_classes = 19
> net = BiSeNet(n_classes=n_classes)
> net.cuda()
> net.load_state_dict(torch.load('./pretrained_checkpoints/79999_iter.pth'))
> net.eval()
> 
> 
> config = dict(img_size=256,id_dim=512,wing_path='pretrained_checkpoints/wing.ckpt')
> nets, nets_ema = build_model(config)
> for name, module in nets.items():
>     nets[name].to('cuda')
>     if ('ema' not in name) and ('fan' not in name) and ('arcface' not in name):
>         print('Initializing %s...' % name)
>         module.apply(utils.he_init)
> for name, module in nets_ema.items():
>     nets_ema[name].to('cuda')
> ckptios = [CheckpointIO('pretrained_checkpoints/', **nets_ema)]
> for ckptio in ckptios:
>     ckptio.load_test('faceswapper.ckpt')
>     
>     
> src_dir = 'data/FF++_Dataset/ff++/raw_003_0.png'
> tar_dir = src_dir.replace('raw_003_0.png', 'raw_000_0.png')
> 
> to_tensor = transforms.Compose([
>         transforms.ToTensor(),
>         transforms.Normalize((0.485, 0.456, 0.406), (0.229, 0.224, 0.225)),
> ])
> 
> img_size = 256
> transform = transforms.Compose([
>         transforms.Resize([img_size, img_size]),
>         transforms.ToTensor(),
>         transforms.Normalize(mean=[0.5, 0.5, 0.5],
>                              std=[0.5, 0.5, 0.5]),
> ])
> transform_seg = transforms.Compose([
>         transforms.Resize([img_size, img_size]),
>         transforms.ToTensor(),
> ])
> 
> def get_mark_parsing(src_img):
>     image = src_img.resize((512, 512), Image.BILINEAR)
>     image = to_tensor(image)
>     image = torch.unsqueeze(image, 0).cuda()
>     out = net(image)[0]
>     
>     parsing = out.squeeze(0).cpu().detach().numpy().argmax(0)
>     
>     mark = parsing.copy()
>     index = np.where(((parsing>0) & (parsing<=6)) | ((parsing>= 10) & (parsing<=13)))
>     mark[index[0], index[1]] = 255
>     index = np.where(mark!=255)
>     mark[index[0], index[1]] = 0
>     mark = Image.fromarray(np.uint8(mark))
>     
>     parsing = Image.fromarray(np.uint8(parsing))
>     return mark, parsing
> 
> source_image = Image.open(src_dir).convert('RGB')
> source_mask_image, source_parsing = get_mark_parsing(source_image)
> 
> tar_image = Image.open(tar_dir).convert('RGB')
> tar_mask_image, tar_parsing = get_mark_parsing(tar_image)
> 
> source_lm_image = transform(source_image)
> source_lm_image = nets_ema.fan.get_landmark_curve(source_lm_image.to('cuda').unsqueeze(0))
> source_lm_image = source_lm_image[0]
> source_lm_image = Image.fromarray(source_lm_image)
> 
> tar_lm_image = transform(tar_image)
> tar_lm_image = nets_ema.fan.get_landmark_curve(tar_lm_image.to('cuda').unsqueeze(0))
> tar_lm_image = tar_lm_image[0]
> tar_lm_image = Image.fromarray(tar_lm_image)
> 
> source_image = transform(source_image)
> source_lm_image = transform(source_lm_image)
> source_mask_image = transform_seg(source_mask_image)
> source_parsing = transform_seg(source_parsing)
> 
> tar_image = transform(tar_image)
> tar_lm_image = transform(tar_lm_image)
> tar_mask_image = transform_seg(tar_mask_image)
> tar_parsing_image = transform_seg(tar_parsing)
> 
> src=source_image.to('cuda')
> tar=tar_image.to('cuda')
> src_lm=source_lm_image.to('cuda')
> tar_lm=tar_lm_image.to('cuda')
> src_mask=(1-source_mask_image).to('cuda')
> tar_mask=(1-tar_mask_image).to('cuda')
> src_parsing=source_parsing.to('cuda')
> tar_parsing=tar_parsing_image.to('cuda')
> 
> srcid_taratt, tarid_srcatt,_,_,_,_ = nets_ema.generator(src.unsqueeze(dim=0), 
>                                                     tar.unsqueeze(dim=0), 
>                                                     src_lm.unsqueeze(dim=0),
>                                                     tar_lm.unsqueeze(dim=0), 
>                                                     src_mask.unsqueeze(dim=0), 
>                                                     tar_mask.unsqueeze(dim=0))
> def denormalize(x):
>     out = (x + 1) / 2
>     return out.clamp_(0, 1)
> x = srcid_taratt[0,:,:,:]
> x = denormalize(x)
> vutils.save_image(x.cpu(), 'test.jpg', nrow=1, padding=0)
> ```

A few issues. In this code src_parsing, src_parsing_image, tar_parsing, tar_parsing_image are never used.

**Comment by winterfell2021** [comment]
@cwalt2014 src_parsing, src_parsing_image, tar_parsing, tar_parsing_image is used for parsing. If u would like to use, please refer to original code.

**Comment by ghost** [comment]
@winterfell2021 No, what I mean is that they are in your code, but you specifically are not using them. Thus you have bugs in your code.
